### ISSUE TITLE
History synchronization

### ISSUE THREAD

**Issue by dag** [fish-shell/fish-shell#825]
The question of whether fish can synchronize history between running shells came up twice on IRC today, and has been asked before as well.

With `fishd` I imagine we have some of the infrastructure for this in place already, so perhaps implementing it wouldn't be too hard.  However, I also think this is something not everyone wants to have all the time, and we also want to avoid adding configuration options.  So the question then is, can we accommodate most needs without configuration?

One idea is to have something like say, `history --synchronize`, and then provide events like "executed a commandline" or perhaps it's enough to do a `function -v history`, and then people who want this can script it.

I don't think that idea is optimal: it's not discoverable (you have to script it) and you can only have either behavior at any given time.

I instead propose making the history wrap around to a synchronized history; that is, if you step or search forward in the history when you are at the bottom (such as is the normal state when on the commandline and you haven't used `history-search-backward` yet), you're acting on a reversed history of commands entered in any shell.  This way, the current behavior of `history-search-backward` is retained and can be used just like we've always use it, but we can _also_ initiate a history search _forwards_ to navigate the universally synchronized history.

Remaining questions:
- Should the synchronized history also include items from the local history?  I think it should, because some people will prefer to always use the synchronized history, and because I think it'll be easier to remember to use the synchronized history when you need it than the opposite.
- What should happen if you do something like `C-p C-n C-n`?  I think this should work just like only doing `C-n`, which is to say "get the latest command used in any shell", and which is in line with the view of these two histories "wrapping around" above and below the commandline.

_(BTW, this idea comes from zsh, which I hear can be made to do something like it, although I haven't actually studied exactly how it behaves.)_

**Comment by milentrifonov** [comment]
@dag The thing is that history is saved after the command completes. What I am doing mostly is to ssh-ing to some of my hosts from one terminal, and then I want to ssh to the last host from another new session, but the last ssh command is not available because I am still logged from the first window and the command is not defect finished and write to history. This bothers me a lot. 
Anyone knows a workaround for this ?

**Comment by faho** [comment]
@milentrifonov:

Try

```fish
function save_history --on-event fish_preexec
    history --save
end
```

(define in your config.fish or a sourced file, not an autoloaded function)

**Comment by milentrifonov** [comment]
@faho Thanks, this actually did the trick.

**Comment by yefim** [comment]
Is there a way to have all my fish sessions share history automatically? I usually have 5 (or more) iTerm tabs open each with a different fish session. I don't want to have to run `history --merge` every time I want to re-run a command from one tab in another. I just want one history. Will the following do what I want?

```` fish
function sync_history --on-event fish_preexec
    history --save
    history --merge
end
````

**Comment by faho** [comment]
@yefim: Yeah, that should work. It might cause a bit of slowdown, though.

**Comment by krader1961** [comment]
I don't think that will do quite what you want. Consider the case where you have two fish sessions A and B. Both have issued the prompt and are waiting for you to enter a command. In A you enter a command. In B you press [up-arrow]. Oops! You haven't yet merged the new history so you won't see the most recently entered command in session A.

**Comment by krader1961** [comment]
Also, as @faho pointed out, that's going to be hideously expensive. Possibly enough to be noticeable.

**Comment by yefim** [comment]
Thank you for the quick response @faho and @krader1961 (and for finding that edge case). So is there a way for me to just have one history across all my sessions without it being so expensive?

**Comment by krader1961** [comment]
> So is there a way for me to just have one history across all my sessions without it being so expensive?

Not at this time. Even if we were to implement this in the core fish code (as opposed to a fish script) it would be very expensive and thus noticeable to the user. Which is why we're unlikely to ever implement this feature.

What I do is define an alias:

```
alias hr 'history --merge'  # read and merge history from disk
```

When I want to have access to the commands I've typed in a different session I just type `hr`. Most of the time I don't want my local command history to be polluted by history from other sessions.

**Comment by ridiculousfish** [comment]
An alternative is to define the up arrow key binding to perform the merge. Then you only merge right before you read history.

**Comment by wprater** [comment]
> An alternative is to define the up arrow key binding to perform the merge. Then you only merge right before you read history.

@ridiculousfish can you show an example?  would this interfere with a up arrow binding to a history search?

**Comment by ridiculousfish** [comment]
Well, it would be something like:

    bind -k up 'history --merge ; up-or-search'

though now that I try it, I see that `history --merge` doesn't properly interleave items - items new to this session are still the most recent in history. That's a separate issue we'd have to fix.

**Comment by ridiculousfish** [comment]
That's me rediscovering #2312

**Comment by bittrance** [comment]
FWIW @ridiculousfish solution does not work for me on 2.3.1 on MacOS X 10.11.5. Instead mapping
```
bind \e\[A 'history --merge ; up-or-search'
```
works fine. I suppose this depends on which binding "wins" when both gets defined (one from default, one from fish_user_key_bindings).

**Comment by faho** [comment]
@bittrance: That probably means your `$TERM` is incorrect. We rely on it to tell us which keys generate a given escape sequence. Run `fish_key_reader` and press UP. What does it print? What's your terminal, what's your "$TERM"?

**Comment by edouard-lopez** [comment]
I'm on the same situation than @bartdag use multiple `fish` instance in `tmux` with different pane/window. 

I asked how to How do I execute a command on pane/window switch? tmux/tmux/issues/666

**Comment by mindreader** [comment]
Could you use some sort of inotify trick on each shell instance to "watch" for changes to the history and merge them after second or two delay?

**Comment by introom** [comment]
so I am using this binding,

# the key seq is \c-; on my terminal.
bind \e"[27;5;59~" "history merge; commandline -f force-repaint"

The shell has the history "echo 3 5", and when I type "echo",
I got the correct hint:
![image](https://user-images.githubusercontent.com/669045/27766882-efb0aab8-5f13-11e7-8c95-3a0a3e026a43.png)

but when I run "echo 3 8" in another fish shell, and I switch back to the original shell and run the above key binding, the hint doesn't get updated to the new "echo 3 8". 
Only after I back-delete a char or use the up/down arrow key or some changes to the command line will that be updated.

**Comment by krader1961** [comment]
@introom, That is the expected behavior. The command reader, highlighter, and suggestion mechanisms "remember" the position in the local history and thus the command that is suggested. Simply doing a history merge won't affect that. The force-repaint isn't intended to change that state.

**Comment by introom** [comment]
@krader1961 any graceful solution in your mind that can update the suggestion?

**Comment by krader1961** [comment]
It's not obvious that behavior is desirable, @introom. But you can achieve it by clearing the command line then reinstating the original typed text. For example:

```
bind \cx "history merge; set -l x (commandline -b); commandline -r ''; commandline -r \$x"
```

Note that you don't even need to force a repaint (which makes it slower). No guarantees that trick will work in the future.

**Comment by farcaller** [comment]
Is there any way to mimic zsh behavior, i.e. call up the last **local** command on the first up-arrow, but then resort to merged history?

**Comment by ridiculousfish** [comment]
Sorry, I can't think of any way to do that.

**Comment by farcaller** [comment]
@ridiculousfish apparently a little hack to `up-or-search` does it:

```fish
function up-or-search -d "Depending on cursor position and current mode, either search backward or move up one line"
    # If we are already in search mode, continue
    if commandline --search-mode
        commandline -f history-search-backward
        return
    end

    # If we are navigating the pager, then up always navigates
    if commandline --paging-mode
        commandline -f up-line
        return
    end

    # We are not already in search mode.
    # If we are on the top line, start search mode,
    # otherwise move up
    set lineno (commandline -L)

    switch $lineno
        case 1
            commandline -f history-search-backward
            history merge # <-- ADDED THIS

        case '*'
            commandline -f up-line
    end
end
```

**Comment by hherman1** [comment]
@ridiculousfish What do you think about making merging on up arrow the default?

**Comment by ridiculousfish** [comment]
I think per session history is the right default.

**Comment by 2m** [comment]
Thank you @farcaller for the workaround! I packaged it up to a plugin for easy usage: https://github.com/2m/fish-history-merge
